File: 0020-valid-parentheses/0020-valid-parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    stack = []

    s.each_char do |char|
        if char == ')' && stack.last == '('
            stack.pop
        elsif char == '}' && stack.last == '{'
            stack.pop
        elsif char == ']' && stack.last == '['
            stack.pop
        else
            stack.push(char)
        end
    end

    !(stack.length > 0)
end

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        stack = []
        
        for elem in s:
            if elem == '(' or elem == '[' or elem == '{':
                stack.append(elem)
            else:
                if len(stack) != 0 and elem == ')' and stack[-1] == '(':
                    stack.pop()
                elif len(stack) != 0 and elem == '}' and stack[-1] == '{':
                    stack.pop()
                elif len(stack) != 0 and elem == ']' and stack[-1] == '[':
                    stack.pop()
                else:
                    return False
        
        return len(stack) == 0